### DISCUSSION TITLE
Best practice to use different devices on I2C bus

### DISCUSSION THREAD

**Discussion by drucknase** [olikraus/u8g2#1749] [General]
Hi,
I have a design with the following key parameters:

- ESP8266 using GPIO4 for I2C SDA and GPIO5 for I2C SCL (standard assignment), 4.7k Pullup at both wires
- connected to the I2C bus 1.3" OLED 128x64 with SH1106 controller and a CO2-sensor Sensirion SCD30
- for the OLED I use the u8x8 library (in platformio lib_deps = olikraus/U8g2@^2.32.7)
- SCD30 needs a I2C clock frequency at 100kHz max. with option clock-stretching
- for SCD30 I use the parkFun SCD30 Arduino Library (in platformio lib_deps = sparkfun/SparkFun SCD30 Arduino Library@^1.0.16). This library use Wire-Library for I2C communication

For the moment I use the following code (code fragment):
`#include <Wire.h>
#include <SparkFun_SCD30_Arduino_Library.h>
#include <U8x8lib.h>

// OLED object
// define OLED i/f
U8X8_SH1106_128X64_NONAME_HW_I2C u8x8(/* reset=*/ U8X8_PIN_NONE); 	  // SSD1306 AZ 1,3" OLED 128x64 

` void setup() {

    // initialize I2C for OLED-display

    u8x8.begin();

    u8x8.setBusClock(120000);

    // Start I2C-Bus for CO2-sensor

    Wire.begin(I2C_SDA,I2C_SCL);

    if (Wire.status() != I2C_OK) Serial.println("Please check I2C wiring");

    // SCD30X I2C Clock Stretching and a  max. 100kHz I2C Clock

    Wire.setClock(120000);

    Wire.setClockStretchLimit(200000L); // CO2-SCD30

} `
Note: the clock value of 120000 gives in reality about 95kHz, which is ok for SCD30.

My question now is: **what is the best practice to initialize the I2C bus in this case?** Should I use SW_I2C for the OLED?
The u8x8.setBusClock() does influence the Wire.setClock(), also the order of initialization makes a difference. 
What about clock stretching for u8x8 lib?

Thank you for some recommendations for the above setup.

**Comment by olikraus** [comment]
Clock stretching could be used by an I2C client, but I do not know any display which uses clock stretching, so it is silently ignored in U8g2.

If your SCD30 lib is based on Wire lib and the display is at the same bus, then you must use HW_I2C for U8g2. In this case both libs will use the same Wire interface (see also the FAQ).

Using u8x8.setBusClock(120000); is probably required (not for U8g2, but for your sensor).
It might be required to call  u8x8.setBusClock() before u8x8.begin(), but if it works, then fine.

> Should I use SW_I2C for the OLED?

No SW_I2C will not work if both devices (display and sensor) share the same bus.

**Reply by drucknase** [reply]
Thank you for the answer and clarification. I will change the order of u8x8.setBusClock() and the u8x8.begin()
